### PR TITLE
PXF: Validation for wrapper options

### PR DESCRIPTION
**Pull request by frankgh**
Validate wrapper options

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: contrib/pxf_fdw/expected/pxf_fdw.out [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ERROR:  protocol is required for PXF foreign-data wrappers
+-- When 'protocol' option is dropped, an error should be given to the user
+-- and the WRAPPER alteration should be aborted.
+ALTER FOREIGN DATA WRAPPER jdbc_pxf_fdw
+    OPTIONS ( DROP protocol );
+ERROR:  protocol is required for PXF foreign-data wrappers
+-- When 'protocol' option is changed to any value, an error should be given
+-- to the user and the WRAPPER alteration should be aborted.
+-- ALTER FOREIGN DATA WRAPPER jdbc_pxf_fdw
+--     OPTIONS ( SET protocol 'non-jdbc' );
```
Comment (denalex) [inline contrib/pxf_fdw/expected/pxf_fdw.out]: if we can't make this work, we should remove these lines before merging the PR.

File: contrib/pxf_fdw/pxf_fdw.c [hunk 2]
```diff
@@ -29,11 +29,12 @@
```
Comment (denalex) [inline contrib/pxf_fdw/pxf_fdw.c]: do we really need all these includes ?

File: contrib/pxf_fdw/pxf_fdw.c [hunk 3]
```diff
@@ -48,17 +49,4 @@ pxf_fdw_handler(PG_FUNCTION_ARGS)
 	// TODO: populate pointers
```
Comment (denalex) [inline contrib/pxf_fdw/pxf_fdw.c]: do we still need to do this ? If not, remove TODO

File: contrib/pxf_fdw/pxf_option.c [hunk 4]
```diff
@@ -0,0 +1,80 @@
+/*
+ * pxf_option.c
+ *		  Foreign-data wrapper for PXF (Platform Extension Framework)
```
Comment (denalex) [inline contrib/pxf_fdw/pxf_option.c]: should we explain what this file is for ?

File: contrib/pxf_fdw/pxf_option.c [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+#include "commands/copy.h"
+#include "commands/defrem.h"
+#include "commands/explain.h"
+#include "optimizer/paths.h"
+#include "optimizer/pathnode.h"
+#include "optimizer/planmain.h"
+#include "parser/parsetree.h"
+#include "optimizer/restrictinfo.h"
+
+static char *const SERVER_OPTION_PROTOCOL = "protocol";
```
Comment (denalex) [inline contrib/pxf_fdw/pxf_option.c]: this is an FDW option, not really SERVER one, should name maybe OPTION_PROTOCOL or FDW_OPTION_PROTOCOL ?

File: contrib/pxf_fdw/pxf_option.c [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+#include "foreign/fdwapi.h"
+#include "foreign/foreign.h"
+#include "commands/copy.h"
+#include "commands/defrem.h"
+#include "commands/explain.h"
+#include "optimizer/paths.h"
+#include "optimizer/pathnode.h"
+#include "optimizer/planmain.h"
+#include "parser/parsetree.h"
+#include "optimizer/restrictinfo.h"
```
Comment (denalex) [inline contrib/pxf_fdw/pxf_option.c]: all these are needed ?

File: contrib/pxf_fdw/pxf_option.c [hunk 7]
```diff
@@ -0,0 +1,80 @@
+/*
```
Comment (denalex) [inline contrib/pxf_fdw/pxf_option.c]: should there be a `pxf_option.h` file ?

File: contrib/pxf_fdw/pxf_option.c [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+		{
+			protocol = defGetString(def);
+		}
+		else
+			other_options = lappend(other_options, def);
+	}
+
+	if (catalog == ForeignDataWrapperRelationId)
+	{
+		if (protocol == NULL || strcmp(protocol, "") == 0)
```
Comment (denalex) [inline contrib/pxf_fdw/pxf_option.c]: what about a string full of whitespace ?

File: contrib/pxf_fdw/pxf_option.c [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+		/*
+		 * Separate out protocol and column-specific options
+		 */
+		if (strcmp(def->defname, SERVER_OPTION_PROTOCOL) == 0)
+		{
+			protocol = defGetString(def);
+		}
+		else
+			other_options = lappend(other_options, def);
```
Comment (denalex) [inline contrib/pxf_fdw/pxf_option.c]: what is the point of copying the options to another list ? nothing else is currently used, right ? We allow other options for FDW, not only `protocol`, correct ?

File: contrib/pxf_fdw/pxf_option.c [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	if (catalog == ForeignDataWrapperRelationId)
+	{
+		if (protocol == NULL || strcmp(protocol, "") == 0)
+			ereport(ERROR,
+			        (errcode(ERRCODE_FDW_DYNAMIC_PARAMETER_VALUE_NEEDED),
+				        errmsg("protocol is required for PXF foreign-data wrappers")));
+	}
+
```
Comment (denalex) [inline contrib/pxf_fdw/pxf_option.c]: what about `else` ? If something else other that FDW is being created, we do not allow protocol, right ? Is it part of another story ?

File: contrib/pxf_fdw/pxf_option.c (hunk 7 above)
Comment (frankgh) [reply contrib/pxf_fdw/pxf_option.c]: not if we contain validation to `pxf_option`

File: contrib/pxf_fdw/pxf_option.c (hunk 8 above)
Comment (frankgh) [reply contrib/pxf_fdw/pxf_option.c]: I thought about that, but I think we can let the backend handle those cases.

File: contrib/pxf_fdw/pxf_option.c (hunk 9 above)
Comment (frankgh) [reply contrib/pxf_fdw/pxf_option.c]: additional checks, I will add the additional check logic

File: contrib/pxf_fdw/expected/pxf_fdw.out (hunk 1 above)
Comment (frankgh) [reply contrib/pxf_fdw/expected/pxf_fdw.out]: yeah, and only superusers can alter FDWs, if they want to shoot themselves in the foot, we will let them

File: contrib/pxf_fdw/pxf_option.c (hunk 10 above)
Comment (frankgh) [reply contrib/pxf_fdw/pxf_option.c]: it's part of another story

File: contrib/pxf_fdw/pxf_option.c (hunk 8 above)
Comment (shivzone) [reply contrib/pxf_fdw/pxf_option.c]: yes i think it makes sense to jsut restrict C side validation for not emptiness check and actual value validation on the server side

Comment (frankgh) [reply contrib/pxf_fdw/pxf_option.c]: We captured this in the story.
> The actual value of the option will not be verified by the C-side, but will be verified later in PXF (in later stories)

**Pull request merged**